### ISSUE TITLE
Redo history lost after adding link to text

### ISSUE THREAD

**Issue by etoledom** [wordpress-mobile/gutenberg-mobile#2091]
**Describe the bug**
Redo is lost after redoing a link on rich-text.

**To Reproduce**
1. Start an empty post.
2. Add a paragraph block and write something.
3. Add formats (Bold, Italic, etc...).
4. Add a link to one of the words **checking `Open in new tab` option.**
5. Add more changes (formats or new words).
6. Undo changes until the link is undone.
7. Try to Redo all the changes again.
8. 💥 The Redo history will be lost.

**Expected behavior**
Redo should work redoing all changes that were undone previously.

**Video**

https://cloudup.com/cOPGvdNJnli

**Smartphone (please complete the following information):**
 - Device: [e.g. iPhone6, Pixel 4]
 - OS: [e.g. iOS 8.1,  Android X]

**Additional context**

**Comment by guarani** [comment]
This is also present on Android, although the behavior is slightly different I think it's the same issue. Also, I noticed that Step 5 from the steps to reproduce can be omitted while still producing the bug.

### Debugging

The bug stems from an extra undo entry being added to undo history in the Redux store right after adding links that set the "Open in new tab option". As an example, suppose a link is added to the letter "two" in the string "One two three". Here's the sequence of events:
1. User types "One two three", selects the "two" portion of the text, taps toolbar link button, bottom sheet is displayed, user fills in link URL (e.g. "www.example.com"), switches **on** "Open in new tab option" and dismisses the bottom sheet. Behind the scenes, an undo entry is correctly added to the Redux store containing the following:
    ```
    "content": "One<a href=\"http://www.example.com\" target=\"_blank\" rel=\"noreferrer noopener\">two</a>three",
    ```
2. When the user subsequently taps anywhere on the line to continue entering text, **an extra undo object is added to the Redux store**. This object contains the following:
    ```
    "content": "One<a rel=\"noreferrer noopener\" href=\"http://www.example.com\" target=\"_blank\">two</a>three",
    ```
    Notice the `rel` attribute has moved from last place to first place in the tag.

Here's a diff of the entry that's added correctly to the undo history (on the left) and the entry (on the right) that's then added incorrectly to the undo history when the user sets the cursor after adding the link. You'll see the only difference is the position of the `rel` attribute inside the `<a>` tag.

<img width="1404" alt="Screen Shot 2020-07-21 at 20 16 39" src="https://user-images.githubusercontent.com/1898325/88128563-415e5e00-cba4-11ea-9e5b-8847f2940fca.png">

What makes me quite sure this is what breaks the redo button functionality as described in this issue, is that if I use the debugger to prevent this erroneous entry from being added, this error disappears. To prevent the entry from being added, I put a breakpoint inside the [undo reducer](https://github.com/WordPress/gutenberg/blob/d86cd5f0b36b3de3034096c09e0aaf7509367c4c/packages/core-data/src/reducer.js#L468) and popped the extra entry off the `nextState` array as soon as it's added.

**Comment by hypest** [comment]
> This is also present on Android, although the behavior is slightly different I think it's the same issue

Hey @guarani, what's the different in Android's case? And, is it possibly related to https://github.com/wordpress-mobile/gutenberg-mobile/issues/2495?

**Comment by hypest** [comment]
This ticket is not worked on at the moment so, I'll clear the assignee field to denote so.

**Comment by fluiddot** [comment]
I reproduced this issue in the latest version of the editor `1.88.0` so it's still present. The undo/redo functionality is part of the writing flow so I think this should be a high priority, hence I'm going to set the High priority label.

**Comment by derekblank** [comment]
Some initial findings:

@guarani's [comment](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2091#issuecomment-662212872) has some very accurate information. I paired with @fluiddot on this for a bit, and we both confirmed that the `rel` attribute shifts places in the link HTML after text has been typed, as noted in the comment and screenshot. There appears to be a state event that is invoked when the `rel` attribute is shifted, which causes the edit reducer to [disrupt the redo history](https://github.com/WordPress/gutenberg/blob/trunk/packages/core-data/src/reducer.js#L523-L558). There are several related issues that describe the broken behavior with undo/redo when using the "Open in new tab" option, so this issue may have a broader scope than is currently known. 

Interestingly, when debugging the same behavior on the web, I noted that the `rel` attribute also shifts within the link HTML after typing, so this behavior is not specific to mobile. When the option to Open in New Tab is selected, the `target="_blank"` and `rel="noreferrer noopener"` attributes are added at the end of the link's HTML. Then, when typing text after the link, the `rel="noreferrer noopener"` jumps to the beginning of the link's HTML, while `target="_blank"` remains at the end: 

https://user-images.githubusercontent.com/643285/233016532-176ee6df-4ead-4867-b449-e1181e55b2cf.mov

My next steps are:

1. Determine if there is a purpose for the `rel` attribute to shift places like in the example.
2. If there is no purpose, determine the _cause_ of where the `rel` attribute shifts in the code (perhaps it is in the shared [createLinkFormat](https://github.com/WordPress/gutenberg/blob/trunk/packages/format-library/src/link/utils.js#L91-L108) util?)
3. If the `createLinkFormat` util is the cause of the `rel` shift and the `rel` shift has no technical purpose, modify the `createLinkFormat` code so the `rel` does not shift.

The scope of the issue seems quite broad. For example, the undo/redo issues have been known within the [manual test cases](https://github.com/wordpress-mobile/test-cases/blob/trunk/test-cases/gutenberg/writing-flow/undo-redo.md) caveats for three years. It may be worth organizing all of the broader undo/redo issues together under a single issue if they are related, which I am happy to start. Open to feedback on how we might better organize this as a broader issue. 🙇 

For reference, here are some of the other undo/redo issues that seem related (some are marked as complete as they are duplicates of each other):

* https://github.com/wordpress-mobile/gutenberg-mobile/issues/2495
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/2091
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1160
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1161
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/3136
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/1158
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/307
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/360

**Comment by derekblank** [comment]
Also, a follow-up on [this question](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2091#issuecomment-662916320):

> What's the different in Android's case? And, is it possibly related to https://github.com/wordpress-mobile/gutenberg-mobile/issues/2495?

It does seem very much related to #2495. However, the difference in my testing on Android is that the redo history is preserved (when it is not on iOS), but on Android it is much more difficult to perform an undo action on a link that has been configured to Open in a New Tab. This aligns with the feedback already captured in #2495. So, while they seem very related, at the moment I do not think they are duplicate issues.

**Comment by fluiddot** [comment]
> Interestingly, when debugging the same behavior on the web, I noted that the `rel` attribute also shifts within the link HTML after typing, so this behavior is not specific to mobile. When the option to Open in New Tab is selected, the `target="_blank"` and `rel="noreferrer noopener"` attributes are added at the end of the link's HTML. Then, when typing text after the link, the `rel="noreferrer noopener"` jumps to the beginning of the link's HTML, while `target="_blank"` remains at the end:

It's interesting that this behavior can also be reproduced in the web version. I tried to reproduce this issue and undo/redo actions work as expected, adding a link to one of the words and checking _Open in new tab_ option doesn't break the undo/redo history. I'm wondering what's the difference in this logic between platforms that only affects the native version. 🤔

**Comment by derekblank** [comment]
> It's interesting that this behavior can also be reproduced in the web version. [...] I'm wondering what's the difference in this logic between platforms that only affects the native version. 🤔

I agree. It's also interesting that there are slight differences between iOS and the similar behavior on Android reported in [#2495](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2495). Given that the web implementation works as expected, this seems to point to Aztec's handling of HTML attributes as the source of these differences between web, iOS, and Android.

**Comment by geriux** [comment]
Hey @derekblank 👋 It looks like this issue was solved in https://github.com/WordPress/gutenberg/pull/50460 if so, can we close this issue? Thanks!

**Comment by derekblank** [comment]
@geriux Yep, closing. 👍
